### ISSUE TITLE
Plotting a model (with model_to_dot) fails if the inputlabels/outputlabels contain brackets

### ISSUE THREAD

**Issue by cBournhonesque** [keras-team/keras#15795]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Macos BigSur
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.7
- Python version: 3.8
- Bazel version (if compiling from source): 
- GPU model and memory:
- Exact command to reproduce:
```
tf.keras.utils.plot_model(model,
                              to_file='model_dir/model.png'),
                              show_shapes=True,
                              show_dtype=True,
                              show_layer_names=True)
```

**Describe the problem**.

When I try to plot a model that contains layers with dictionary inputs, I get an error in the 
`model_to_dot` function in `vis_utils.py`
that says: `Error: invalid label format`.
This error comes from having an invalid graphviz label name (defined here: https://github.com/keras-team/keras/blob/master/keras/utils/vis_utils.py#L293)
My input shape is a dictionary, and is in the form: `{'a': (None, 1), 'b': (None, 2)}`. 
If I call `plot_model` with `show_shapes=True`, then the shape will be added to the label name for graphbiz.
The problem is that the brackets in the `inputlabels` and `outputlabels` need to be escaped so that the node's label can be interpreted correctly by graphviz. (otherwise, graphviz interprets it as a nested label: https://graphviz.org/doc/info/shapes.html#record)

I fixed the issue by adding:
```
inputlabels = inputlabels.replace('{', '\{')
inputlabels = inputlabels.replace('}', '\}')
outputlabels = outputlabels.replace('{', '\{')
outputlabels = outputlabels.replace('}', '\}')
```
before https://github.com/keras-team/keras/blob/master/keras/utils/vis_utils.py#L293
but there must be a more elegant way of fixing this.

**Describe the current behavior**.
Plotting the model fails.

**Describe the expected behavior**.
The model gets plotted correctly

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):
Maybe we could have a list of characters to escape ('{', '}', '|', etc.), and escape all of these at once in the label to be plotted?

**Comment by jvishnuvardhan** [comment]
@cBournhonesque Can you please share a simple standalone code to reproduce the issue? Thanks!

**Comment by cBournhonesque** [comment]
You were right to ask!
It looks like when the input is a dictionary, the brackets are already removed, so the graph works:
![image](https://user-images.githubusercontent.com/8112632/146255072-a999c833-5940-4350-ba2e-81ec5cfece8b.png)

The error comes when the input is a tuple of Dicts, that's where you have brackets in the label names:

Example code:
```
from typing import Dict, Tuple
import tensorflow as tf

class DictLayer(tf.keras.layers.Layer):
    def call(self, inputs: Tuple[tf.Tensor, Dict[str, tf.Tensor]]) -> tf.Tensor:
        tensor_input, dict_input = inputs
        return tf.concat(list(dict_input.values()), axis=1)

inputs={"a": tf.keras.Input(name="a", shape=(1), dtype=tf.float32),
        "b": tf.keras.Input(name="b", shape=(1), dtype=tf.float32)}
outputs=DictLayer()((inputs["a"], inputs))
model = tf.keras.Model(
    inputs=inputs,
    outputs=outputs,
)
tf.keras.utils.plot_model(model,
                          to_file='model.png',
                          show_shapes=True,
                          show_dtype=True,
                          show_layer_names=True)
```

Error: `Error: bad label format dict_layer|DictLayer|float32\\n|{input:|output:}|{{((None, 1), {'a': (None, 1), 'b': (None, 1)})}|{(None, 2)}}\n"`

**Comment by cBournhonesque** [comment]
Please do not close the issue.

**Comment by stevetracvc** [comment]
It also happens when you show activations and use something like PReLU:

```
b"Error: bad label format layer_31|{layer_31: Dense|{'class_name': 'PReLU', 'config': {'name': 'p_re_lu', 'trainable': True, 'dtype': {'class_name': 'PolicyV1', 'config': {'name': 'mixed_float16'}}, 'alpha_initializer': {'class_name': 'Zeros', 'config': {}}, 'alpha_regularizer': None, 'alpha_constraint': None, 'shared_axes': None}}}|float32\\n|{input:|output:}|{{(None, 1440)}|{(None, 2848)}}\n"
```
The reason being, it calls `activations.serialize(layer.activation)` which outputs the results of get_config() (a dict) for a more complex activation, whereas if the activation is just "relu" then there is no get_config() so the serialize function just spits back the string.

**Comment by jvishnuvardhan** [comment]
@cBournhonesque This was resolved with your PR. I am closing this issue as this was resolved. Please feel free to reopen If I am mistaken. Thanks!

**Comment by mattdangerw** [comment]
This looks like it is breaking things as of tensorflow 2.10 with nested models. Attached a repro...
https://colab.sandbox.google.com/gist/mattdangerw/22a9dcb0319875f7039e6335ac5614ec/plot-model-bug.ipynb

Any ideas anyone?

**Comment by saberkun** [comment]
bad label format error happens to a lot of tf model garden models and broke colabs. It is a regression and still fails in tf 2.11

**Comment by jayanta-banik** [comment]
there was an error in 
/usr/local/lib/python3.9/dist-packages/keras/utils/vis_utils.py

line 278

```
def format_shape(shape):
                return (
                    str(shape)
                    .replace(str(None), "None")
                    .replace("{", "/{")  # change this to \{
                    .replace("}", "/}") # change this to \}
                )
```

PS im still new to reporting error. 

Edit: it appears to be fixed now

**Comment by tilakrayal** [comment]
@cBournhonesque,
I tried to execute the mentioned code on tf-nightly(2.15.0-dev20230922) and it was executed without any issue/error and the model was also plotted. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/b5b504ebfece8c1f419a9c3a293c5f3b/untitled1372.ipynb) and let us know if it serves the purpose. Thank you!
